src/distribution_spatial.cpp

```diff
@@ -180,6 +185,84 @@ Position SphericalIndependent::sample(uint64_t* seed) const
   return {x, y, z};
 }
 
+//==============================================================================
+// MeshIndependent implementation
+//==============================================================================
+
+// Loosely adapted Patrick's code shown here
+MeshIndependent::MeshIndependent(pugi::xml_node node)
+{
+  // No in-tet distributions implemented, could include distributions for the barycentric coords
+  // Read in unstructured mesh from mesh_id value
+  int32_t mesh_id = std::stoi(get_node_value(node, "mesh_id"));
+  mesh_map_idx_ = model::mesh_map.at(mesh_id);
+  const auto& mesh_ptr = model::meshes[mesh_map_idx_];
+  
+  // Check whether mesh pointer points to an unstructured mesh
+  umesh_ptr_ = dynamic_cast<UnstructuredMesh*>(mesh_ptr.get());
+  if (!umesh_ptr_) {fatal_error("Mesh passed to spatial distribution is not an unstructured mesh object"); }
+
+  // Create CDF based on weighting scheme
+  tot_bins_ = umesh_ptr_->n_bins();
+  // float weights[tot_bins_] = {};
+  std::vector<double> weights = {};
+  weights.resize(tot_bins_);
+  double temp_total_weight = 0.0;
+  mesh_CDF_.resize(tot_bins_+1);
+  mesh_CDF_[0] = {0.0};
+  total_weight_ = 0.0;
+  mesh_weights_.resize(get_node_array<double>(node, "weights_from_file").size());
+
+  // Create cdfs for sampling for an element over a mesh
+  // Volume scheme is weighted by the volume of each tet
+  // File scheme is weighted by an array given in the xml file
+
+  if (check_for_node(node, "elem_weight_scheme")) {
+    sample_scheme_ = get_node_value(node, "elem_weight_scheme");
+  } else {
+      fatal_error("No weighting scheme was specified for element sampling over the mesh");
+  }
+
+  if (sample_scheme_ == "volume"){
+    for (int i = 0; i<tot_bins_; i++){
+      weights[i] = umesh_ptr_->volume(i);
+    }
+  } else if (sample_scheme_ == "file"){
+
+    if (check_for_node(node, "weights_from_file")) {
+      mesh_weights_ = get_node_array<double>(node, "weights_from_file");
+      if (mesh_weights_.size() != tot_bins_){
+        write_message("The size of the weights array from the xml file does not equal the number of elements in the mesh.");
+        // fatal_error("The size of the weights array from the xml file does not equal the number of elements in the mesh.");
+      }
+      for (int i = 0; i < tot_bins_; i++){
+        weights[i] = mesh_weights_[i];
+      } 
+    } else {
+        fatal_error("No weights given in xml file");
+    }
+  } else{
+    fatal_error("Type of element weighting scheme provided is not in the supported types (volume or file)");
+  }
```

Comment:
This is where I think we can simplify how the schemes are handled:

```cpp
// if a set of weights is present in the XML start, with it
if (check_for_node(node, "weights_from_file")) {
  mesh_weights_ = get_node_array<double>(node, "weights_from_file");
  // check that the number of provided weights matches the number of mesh elements
  if (mesh_weights_.size() != tot_bins_) fatal_error("Number of provided weights does not match the number of mesh bins");
} else {
  mesh_weights_.resize(tot_bins_);
  std::fill(mesh_weights_.begin(), mesh_weights_.end(), 1.0);
}

// adjust the weights by volume if requested
bool volume_weighted = true;
if (check_for_node(node, "volume_weighted")) 
  bool volume_weighted = get_node_value_bool(node, "volume_weighted");

if (volume_weighted) {
  for (int i = 0; i < tot_bins_; i++) {
    mesh_weights_[i] *= umesh_ptr_->volume(i);
  }
}
```

This setup would require some small changes to the inputs on the Python side of course. Our current schemes would correspond to the following settings on the `MeshIndependent` class:

  - "volume": no weights provided in the XML, `volume_weighting` is `True`
  - "file" : read weights from XML file, `volume_weighting` set to `False`
 
And going forward to the other schemes we're looking to support:
  - "material source" : read weights (specific activities) in from file, `volume_weighting` set to `True`
  - "uniform" (for testing) : no weights provided, `volume_weighting` set to `False`

I think this makes the code much simpler. I'd be fine getting rid of the schemes altogether, but I can see how that might make the API more confusing to new users. To address that, I could see us keeping the schemes in the Python API and doing some checking on the C++ side based on the scheme selected. For example, if the scheme is "material source" then we would ensure that weights are present in the XML file.

What do you think?

Reply:
I'm still not sure which way is better. As you stated, it may be easier for users to just specify a scheme rather than setting variables to true/false. Since there are only about 4 total schemes (material_source, file, volume, maybe uniform) I think code complexity will be pretty similar either way, while those changes may make it easier to control volume weighting. One reason I chose this scheme is I thought that there may eventually be a material source that does not rely on the outside xml file, instead the activity/materials are tied to the elements themselves and they would be looked up similar to how volume is now. Is this a possibility or will it always be exported to an external file?

Reply:
It's possible that we'd set this using the C++ API, but we'd be doing that with a different constructor than the one taking in an XML node, so the code we're writing here should assume we're always getting that information from file.

I think we can get the best of both by providing a scheme on the `MeshIndependent` class in the Python API to make it easier on users. On the C++ side, we can perform error checking based on the specified scheme as needed, making sure that weights are present as need and the volume weighting is applied as needed. Then we can set the initial weights and apply volume weighting as prescribed above. Sound good?